feat: circuit simulator for Ultra and GoblinUltra verifiers

codygunton opened the pull request:
# Description
This PR introduces the `CircuitSimulator` repurposed specifically to simulate native verifiers via the recursive verifiers, currently specialised for bn254. Currently the simulator is able to simulate the `UltraVerifier` and `GoblinUltraVerifier`.

Resolves https://github.com/AztecProtocol/barretenberg/issues/944.

We also added benchmark to track the slowdown incurred by simulating native verifiers: 

```
---------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                                                             Time             CPU   Iterations
---------------------------------------------------------------------------------------------------------------------------------------
SimulatorFixture<bb::GoblinUltraRecursiveFlavor_<bb::CircuitSimulatorBN254>>/GoblinSimulated       15.2 ms         15.2 ms           46
SimulatorFixture<bb::UltraRecursiveFlavor_<bb::CircuitSimulatorBN254>>/UltraSimulated              13.2 ms         13.2 ms           53
SimulatorFixture<bb::GoblinUltraRecursiveFlavor_<bb::CircuitSimulatorBN254>>/GoblinNative          9.52 ms         9.52 ms           73
SimulatorFixture<bb::UltraRecursiveFlavor_<bb::CircuitSimulatorBN254>>/UltraNative                 8.34 ms         8.34 ms           84
```
The PR also adds several TODOs and linked issues from when @codygunton was working on making a more generic circuit simulator for the stdlib in the context of Aztec circuits in C++.

https://github.com/AztecProtocol/barretenberg/issues/677 Simulating biggfield assert equal
https://github.com/AztecProtocol/barretenberg/issues/676 Should CircuitSimulator be a base class? UltraCircuitSimulator?
https://github.com/AztecProtocol/barretenberg/issues/672 Interfaces for recursion, folding
https://github.com/AztecProtocol/barretenberg/issues/671 One Pedersen
https://github.com/AztecProtocol/barretenberg/issues/670 Simulator Pedersen fragility
https://github.com/AztecProtocol/barretenberg/issues/669 `safe_uint` subtraction
https://github.com/AztecProtocol/barretenberg/issues/668 Extend simulator functionality
https://github.com/AztecProtocol/barretenberg/issues/667 Simulator execution of LogicTest.LargeOperands
https://github.com/AztecProtocol/barretenberg/issues/666 Fix simulator issues in array tests 
https://github.com/AztecProtocol/barretenberg/issues/665 Simulator should satisfy `HasPlookup`(?)
https://github.com/AztecProtocol/barretenberg/issues/664 Possible missing constraint in biggroup `compute_wanf`
https://github.com/AztecProtocol/barretenberg/issues/663 Simulate biggroup `batch_mul` in general #663
https://github.com/AztecProtocol/barretenberg/issues/662 Simulating stdlib keccak (possible bug?)
https://github.com/AztecProtocol/barretenberg/issues/660 `nullptr` context in standard library
https://github.com/AztecProtocol/barretenberg/issues/659 Schnorr and ECDSA extended to non-string arguments
https://github.com/AztecProtocol/barretenberg/issues/563 `MerkleHashType` and `CommitmentType` don't belong in circuit constructor
https://github.com/AztecProtocol/barretenberg/issues/562 `CircuitType` in verification key? Something else?
https://github.com/AztecProtocol/barretenberg/issues/426 `HashType` and `CommitmentType` specification

ludamad committed:
feat: barretenberg PR migrate tool

codygunton committed:
Simulator

codygunton committed:
Reinstate asserts.

codygunton committed:
Give up and skip test.

codygunton committed:
Manually pick non-B. changes.

codygunton committed:
Add one high generator.

codygunton committed:
Merge branch 'master' into cg/simulate-spike-bb-subrepo

codygunton committed:
Reset files I didn't touch

ludamad commented:
Hmm, I'll have to look into why the script undid changes from master. Might be due to the use of `--force`

ludamad committed:
fix: merge and ci workarounds

ludamad committed:
Merge branch 'master' into cg/simulate-spike-bb-subrepo

ludamad committed:
Merge branch 'master' into cg/simulate-spike-bb-subrepo

codygunton committed:
Compare Pedersen benchmarks.

Benchmark                                              Time             CPU      Time Old      Time New       CPU Old       CPU New
-----------------------------------------------------------------------------------------------------------------------------------
pedersen_compress_pair/10/repeats:1                 +0.2210         +0.2209         77876         95085         77887         95095
pedersen_compress_array/10/repeats:1                +0.0200         +0.0199        426862        435396        426919        435432
OVERALL_GEOMEAN                                     +0.1160         +0.1159             0             0             0             0

codygunton committed:
blake3s

codygunton committed:
Ecdsa / secp256k1

codygunton committed:
Biggroup batch_mul

Benchmark                                              Time             CPU      Time Old      Time New       CPU Old       CPU New
-----------------------------------------------------------------------------------------------------------------------------------
pedersen_compress_pair/10/repeats:1                 +0.2269         +0.2279         77967         95661         77983         95753
pedersen_compress_array/10/repeats:1                +0.0243         +0.0244        427006        437372        427089        437513
blake3s/10/repeats:1                              +214.3972       +214.2077           497        107101           498        107236
ecdsa/10/repeats:1                                +427.5370       +427.5895        238771     102322066        238739     102321059
biggroup_batch_mul/10/repeats:1                   +330.4509       +330.4886        964755     319768833        964633     319764957
OVERALL_GEOMEAN                                    +31.8835        +31.8851             0             0             0             0

codygunton committed:
Speed up biggroup batch mul.

Benchmark                                              Time             CPU      Time Old      Time New       CPU Old       CPU New
-----------------------------------------------------------------------------------------------------------------------------------
pedersen_compress_pair/10/repeats:1                 +0.2192         +0.2193         78469         95669         78476         95684
pedersen_compress_array/10/repeats:1                +0.0157         +0.0156        428728        435478        428828        435532
blake3s/10/repeats:1                              +211.0444       +210.5690           499        105862           500        105876
ecdsa/10/repeats:1                                +426.5165       +426.5226        239377     102337525        239360     102331832
biggroup_batch_mul/10/repeats:1                     +0.2060         +0.2059        968405       1167908        968260       1167661
OVERALL_GEOMEAN                                     +9.6247         +9.6197             0             0             0             0

codygunton committed:
Merge remote-tracking branch 'origin/master' into cg/simulate-spike-bb-subrepo

codygunton committed:
Changes missed in merge commit.

codygunton committed:
Try to satisfy GCC

codygunton committed:
Tweak script output.

codygunton committed:
Speed up ECDSA simulation

Benchmark                                              Time             CPU      Time Old      Time New       CPU Old       CPU New
-----------------------------------------------------------------------------------------------------------------------------------
pedersen_compress_pair/10/repeats:1                 +0.2185         +0.2183         78085         95143         78097         95150
pedersen_compress_array/10/repeats:1                +0.0161         +0.0161        429360        436272        429403        436334
blake3s/10/repeats:1                              +211.5298       +210.6946           499        106157           502        106166
ecdsa/10/repeats:1                                  +0.0156         +0.0155        238213        241933        238196        241895
biggroup_batch_mul/10/repeats:1                     +0.1987         +0.1987        963210       1154583        963031       1154370
OVERALL_GEOMEAN                                     +2.1704         +2.1679             0             0             0             0
%

codygunton committed:
Add missing brace

codygunton committed:
Use derived context.

codygunton committed:
Set context so merkel test will run.

codygunton committed:
Fix script comments

codygunton committed:
TODO(https://github.com/AztecProtocol/barretenberg/issues/659)

codygunton commented on circuits/cpp/barretenberg/cpp/src/barretenberg/stdlib/primitives/byte_array/byte_array.cpp:
```diff
@@ -351,9 +351,9 @@ typename byte_array<Composer>::byte_slice byte_array<Composer>::split_byte(const
     if (byte.is_constant()) {
         field_t<Composer> low(context, low_value);
-        field_t<Composer> high(context, high_value);
+        field_t<Composer> shifted_high(context, high_value * (uint64_t(1) << (8ULL - num_high_bits)));
```
This was a bug in circuit construction.

codygunton commented on circuits/cpp/barretenberg/cpp/src/barretenberg/stdlib/primitives/bit_array/bit_array.test.cpp:
```diff
@@ -50,10 +52,8 @@ TYPED_TEST(BitArrayTest, test_uint32_input_output_consistency)
-        static_cast<uint32_t>(composer.get_variable(result[0].get_witness_index()).from_montgomery_form().data[0]);
-    auto b_result =
-        static_cast<uint32_t>(composer.get_variable(result[1].get_witness_index()).from_montgomery_form().data[0]);
+    auto a_result = static_cast<uint32_t>(result[0].get_value());
```
With the simulator we can't get variable values, but anyway, the stdlib tests should not access builder functions.

codygunton commented on circuits/cpp/barretenberg/cpp/src/barretenberg/stdlib/primitives/biggroup/biggroup_nafs.hpp:
```diff
@@ -519,7 +526,10 @@ std::vector<bool_t<C>> element<C, Fq, Fr, G>::compute_naf(const Fr& scalar, cons
+            if constexpr (IsSimulator<C>) {
+                ctx->create_range_constraint(
+                    bit.get_value(), 1, "biggroup_nafs: compute_naf extracted too many bits in next_entry case");
+            } else if constexpr (HasPlookup<C>) {
```
Cf https://github.com/AztecProtocol/barretenberg/issues/665

codygunton commented on circuits/cpp/barretenberg/cpp/src/barretenberg/stdlib/primitives/circuit_builders/circuit_builders.hpp:
```diff
@@ -3,17 +3,24 @@
 
-template <typename T> concept HasPlookup = proof_system::IsAnyOf<T, proof_system::UltraCircuitBuilder>;
+template <typename T>
+concept HasPlookup =
```
I don't know the best place to put this or `IsSimulator`

codygunton commented on circuits/cpp/barretenberg/cpp/src/barretenberg/stdlib/primitives/circuit_builders/circuit_builders.hpp:
```diff
@@ -3,17 +3,24 @@
+concept HasPlookup =
+    proof_system::IsAnyOf<T, proof_system::UltraCircuitBuilder /* , proof_system::CircuitSimulatorBN254 */>;
+
+#define INSTANTIATE_STDLIB_SIMULATOR_METHOD(stdlib_method) template stdlib_method(proof_system::CircuitSimulatorBN254);
```
As we discussed, these macros were just made to allow me to work incrementally. They should go away (perhaps that's already possible?).

codygunton commented on circuits/cpp/barretenberg/cpp/src/barretenberg/stdlib/primitives/curves/bn254.hpp:
```diff
@@ -20,17 +20,17 @@ template <typename CircuitBuilder> struct bn254 {
-    typedef byte_array<CircuitBuilder> byte_array_ct;
-    typedef bool_t<CircuitBuilder> bool_ct;
-    typedef stdlib::uint32<CircuitBuilder> uint32_ct;
+    using witness_ct = witness_t<CircuitBuilder> ;
```
Couldn't result. I just find the `typedef`s so much harder to read...

codygunton commented on circuits/cpp/barretenberg/cpp/src/barretenberg/stdlib/primitives/field/field.test.cpp:
```diff
@@ -120,74 +127,92 @@ template <typename Composer> class stdlib_field : public testing::Test {
+        }
     }
 
-    static void test_add_mul_with_constants()
-    {
-        Composer composer = Composer();
-        auto gates_before = composer.get_num_gates();
-        uint64_t expected = fidget(composer);
-        auto gates_after = composer.get_num_gates();
-        EXPECT_EQ(composer.get_variable(composer.w_o[gates_after - 1]), fr(expected));
-        info("Number of gates added", gates_after - gates_before);
-        bool result = composer.check_circuit();
-        EXPECT_EQ(result, true);
-    }
```
This isnt' really a test of field logic, rather it's a test of circuit builder logic.

codygunton commented on circuits/cpp/barretenberg/cpp/src/barretenberg/stdlib/primitives/uint/uint.hpp:
```diff
@@ -189,19 +189,19 @@ template <typename T, typename w> inline std::ostream& operator<<(std::ostream&
 
 template <typename ComposerContext>
-using uint8 = typename std::conditional<HasPlookup<ComposerContext>,
+using uint8 = typename std::conditional<HasPlookup<ComposerContext> || IsSimulator<ComposerContext>,
```
This has been icky for a long time. There is surely a better way.

codygunton commented on circuits/cpp/barretenberg/cpp/src/barretenberg/stdlib/primitives/uint/uint.test.cpp:
```diff
@@ -1849,7 +1850,11 @@ TYPED_TEST(stdlib_uint, test_divide_special)
 TYPED_TEST(stdlib_uint, div_remainder_constraint)
 {
-    TestFixture::div_remainder_constraint();
+    if constexpr (IsSimulator<CircuitSimulatorBN254>) {
```
Probably the right thing to do is not compile tests we don't expect to pass, so that we can reserve "skip"s for tests we want to remind ourselves are being skipped for whatever reason.

codygunton commented on circuits/cpp/barretenberg/cpp/src/barretenberg/stdlib/recursion/aggregation_state/aggregation_state.hpp:
```diff
@@ -70,8 +70,6 @@ template <typename Curve> struct aggregation_state {
 
         auto* context = P0.get_context();
 
-        context->check_circuit();
```
This was accidentally merged by me a couple of months ago and should go away.

codygunton committed:
Cleanup and issue issues.

codygunton committed:
Remove unneeded function.

codygunton committed:
Reinstate missing recursion check.

codygunton committed:
Clean up and add comments.

ludamad committed:
Intentionally do a bad merge.

ludamad committed:
fix e2e-escrow

ludamad0 committed:
Merge remote-tracking branch 'origin/master' into cg/simulate-spike-bb-subrepo

ludamad0 committed:
Merge remote-tracking branch 'origin/master' into cg/simulate-spike-bb-subrepo

ludamad committed:
chore: merge master into simulate feature branch (#1848)

ludamad0 committed:
Merge

ludamad0 committed:
Rm bb

ludamad0 committed:
git mv bb

ludamad0 committed:
Rm bb/build-system

ludamad0 committed:
Merge commit '404ec34d38e1a9c3fbe7a3cdb6e88c28f62f72e4' into cg/simulate-spike-bb-subrepo

ludamad0 committed:
Merge remote-tracking branch 'origin/master' into cg/simulate-spike-bb-subrepo

ludamad0 committed:
Fix merge markers

ludamad0 committed:
Format

ludamad0 committed:
Reverts

ludamad commented:
Currently not planned. Work can be resurrected later if needed.

Commit:
Merge branch 'master' into cg/simulate-spike-bb-subrepo

Commit:
fine till here

Commit:
experiments with ultra verifier

Commit:
unify honk recursive verifiers

Commit:
Merge remote-tracking branch 'origin/master' into cg/simulate-spike-bb-subrepo

Commit:
yay

Commit:
stuff

Commit:
add benchmark

Commit:
more cleanup

Commit:
Merge remote-tracking branch 'origin/master' into cg/simulate-spike-bb-subrepo

maramihali commented on barretenberg/cpp/src/barretenberg/benchmark/simulator_bench/simulator.bench.cpp:
```diff
@@ -0,0 +1,126 @@
+    }
+};
+
+BENCHMARK_TEMPLATE_F(SimulatorFixture, GoblinNative, bb::GoblinUltraRecursiveFlavor_<bb::CircuitSimulatorBN254>)
```
This and UltraNative are basically the same code (ditto for the simulated ones) but I haven't managed to instantiate with multiple templates, not even sure it's supported. Very much open to suggestions

maramihali commented on barretenberg/README.md:
````diff
@@ -85,7 +85,7 @@ Or install from a package manager, on Ubuntu:
 ```
 
-> Note: on a fresh Ubuntu Kinetic installation, installing OpenMP from source yields a `Could NOT find OpenMP_C (missing: OpenMP_omp_LIBRARY) (found version "5.0")` error when trying to build Barretenberg. Installing from apt worked fine.
+> Note: on a fresh Ubuntu lunar installation, installing OpenMP from source yields a `Could NOT find OpenMP_C (missing: OpenMP_omp_LIBRARY) (found version "5.0")` error when trying to build Barretenberg. Installing from apt worked fine.
````
was this change intended?

Commit:
more cleanup

Commit:
reenable check, more cleanup

Commit:
cleanup benchmark

Commit:
Merge remote-tracking branch 'origin/master' into cg/simulate-spike-bb-subrepo

Commit:
fix gcc and test

ledwards2225 commented on barretenberg/cpp/src/barretenberg/stdlib/primitives/field/field.test.cpp:
```diff
@@ -31,11 +31,12 @@ template <typename Builder> class stdlib_field : public testing::Test {
             a = c;
             c = a + b;
         }
+        return c;
```
I'm surprised this even builds..

ledwards2225 commented on barretenberg/cpp/src/barretenberg/stdlib/primitives/field/field.test.cpp:
```diff
@@ -117,76 +126,97 @@ template <typename Builder> class stdlib_field : public testing::Test {
+        }
     }
 
+    // static void test_add_mul_with_constants()
```
What's going on here?

maramihali replied:
forgot to uncomment

ledwards2225 commented on barretenberg/cpp/src/barretenberg/stdlib/primitives/field/field.test.cpp:
```diff
@@ -249,19 +279,19 @@ template <typename Builder> class stdlib_field : public testing::Test {
-        auto& block = builder.blocks.arithmetic;
-        EXPECT_EQ(builder.get_variable(block.w_l()[block.size() - 1]), fr(4181));
-        EXPECT_EQ(gates_after - gates_before, 18UL);
+    // static void test_field_fibbonaci()
```
?

ledwards2225 commented on barretenberg/cpp/src/barretenberg/stdlib/primitives/field/field.test.cpp:
```diff
@@ -935,10 +984,10 @@ TYPED_TEST(stdlib_field, test_prefix_increment)
-{
-    TestFixture::test_field_fibbonaci();
-}
+// TYPED_TEST(stdlib_field, test_field_fibbonaci)
```
?

ledwards2225 commented on barretenberg/cpp/src/barretenberg/stdlib/primitives/plookup/plookup.test.cpp:
```diff
@@ -14,15 +14,15 @@ using namespace bb;
 
 // Defining ultra-specific types for local testing.
-using Builder = UltraCircuitBuilder;
+using Builder = UltraCircuitBuilder; // SIMULATOOOOR
```
?

maramihali replied:
was trying to figure out what should be simulated initially while bringing this PR to master

ledwards2225 reviewed (commented):
It seems like there's still a bit of cleanup/clarification to be done here. There are several changes that don't seem to be related to the Simulator work and some disabled tests.

My honest opinion is that this is simply not the right thing for what we need. It was a good idea when we thought we were writing ALL circuits in C++. But we're adding complexity to deep stdlib code (along with many conditionals) in order to avoid 3 or 4 conditionals in verifier code? I also don't love the idea of having something that purports to be a circuit simulator that only has limited capacity to mimic real builder behavior. That said, I think I expressed this opinion before we picked this work back up so if you and @Cody think this is the right move then I won't argue.

Commit:
more cleanup

maramihali commented on barretenberg/cpp/src/barretenberg/stdlib/primitives/bit_array/bit_array.test.cpp:
```diff
@@ -46,10 +46,9 @@ TYPED_TEST(BitArrayTest, test_uint32_input_output_consistency)
-        static_cast<uint32_t>(builder.get_variable(result[0].get_witness_index()).from_montgomery_form().data[0]);
-    auto b_result =
-        static_cast<uint32_t>(builder.get_variable(result[1].get_witness_index()).from_montgomery_form().data[0]);
+    // Maybe problem
```
no it's just stale note as I was merging and fixing linking issues

Commit:
even more cleanup

maramihali commented on barretenberg/cpp/src/barretenberg/stdlib/honk_recursion/verifier/verifier.test.cpp:
```diff
@@ -197,7 +197,11 @@ template <typename RecursiveFlavor> class RecursiveVerifierTest : public testing
 
         // Check 3: Construct and verify a proof of the recursive verifier circuit
-        {
+        if constexpr (!IsSimulator<OuterBuilder>) {
```
Think it probably can, I will have an attempt

Commit:
delete some more unnecessary stuff

maramihali commented on barretenberg/cpp/src/barretenberg/stdlib/plonk_recursion/aggregation_state/aggregation_state.hpp:
```diff
@@ -30,12 +30,14 @@ template <typename Curve> struct aggregation_state {
      * @brief TODO(@dbanks12 please migrate A3 circuits to using `assign_object_to_proof_outputs`. Much safer to not
      * independently track `proof_witness_indices` and whether object has been assigned to public inputs)
-     *
+     * @deprecated
```
no we cant yet it's still used by plonk recursive verifier will revert this line

maramihali commented on barretenberg/cpp/src/barretenberg/ecc/scalar_multiplication/scalar_multiplication.cpp:
```diff
@@ -220,7 +220,11 @@ void compute_wnaf_states(uint64_t* point_schedule,
 
     parallel_for(num_threads, [&](size_t i) {
-        Fr T0;
+        // NOTE: to appease GCC array-bounds checks, we need an extra Fr at the end of 'T0'
```
i think it can be reverted

Commit:
more cleanup

maramihali commented on barretenberg/cpp/src/barretenberg/stdlib/primitives/byte_array/byte_array.test.cpp:
```diff
@@ -123,6 +123,22 @@ TYPED_TEST(ByteArrayTest, set_bit)
     STDLIB_TYPE_ALIASES
     auto builder = Builder();
 
+    // byte_array_ct arr(&composer, std::vector<uint8_t>{ 0x01, 0x02, 0x03, 0x04 });
```
deleting

Commit:
revert some formatting

Commit:
remove unnecessary fields in simulator

codygunton commented on barretenberg/cpp/src/barretenberg/stdlib/hash/keccak/keccak.cpp:
```diff
@@ -814,14 +814,21 @@ stdlib::byte_array<Builder> keccak<Builder>::hash(byte_array_ct& input, const ui
-        // if buffer is constant compute hash and return w/o creating constraints
-        byte_array_ct output(nullptr, 32);
+    const auto constant_case = [&] { // if buffer is constant, compute hash and return w/o creating constraints
+        byte_array_ct output(nullptr, static_cast<uint32_t>(num_bytes.get_value() >> 1));
```
I don't recall adding this, but it has been a long time. Maybe it's an artifact of the merge? Does resetting to 32 cause tests to break? If not I would revert to the magic number.

codygunton commented on barretenberg/cpp/src/barretenberg/stdlib/hash/keccak/keccak.test.cpp:
```diff
@@ -196,6 +201,8 @@ TEST(stdlib_keccak, test_single_block)
 
 TEST(stdlib_keccak, test_double_block)
 {
+    GTEST_SKIP() << "Bug in constant case?";
```
Would you please add issues for these and file them appropriately on the kanban?

codygunton commented on barretenberg/cpp/src/barretenberg/stdlib/primitives/byte_array/byte_array.cpp:
```diff
@@ -350,9 +350,9 @@ typename byte_array<Builder>::byte_slice byte_array<Builder>::split_byte(const s
     if (byte.is_constant()) {
         field_t<Builder> low(context, low_value);
-        field_t<Builder> high(context, high_value);
+        field_t<Builder> shifted_high(context, high_value * (uint64_t(1) << (8ULL - num_high_bits)));
```
Yeah IIRC this was a bug fix in a code path that wasn't tested before.

codygunton reviewed (commented):
It's a huge PR that I started a long time ago that has been resurrected and looked at by three people. It still feels a bit like a mess, but it also seems to get the job done correctly as the simulator is tested pretty thoroughly in the stdlib primitives tests.

Commit:
follup up after cody's comments

Commit:
Merge remote-tracking branch 'origin/master' into cg/simulate-spike-bb-subrepo

ledwards2225 reviewed (approved):
Rubber stamp!

Commit:
Merge remote-tracking branch 'origin/master' into cg/simulate-spike-bb-subrepo

Commit:
Merge branch 'master' into cg/simulate-spike-bb-subrepo
